benchmarks/bvh_driver/benchmark_registration.hpp

```diff
@@ -297,6 +310,52 @@ void BM_radius_callback_search(benchmark::State &state, Spec const &spec)
       spec.n_queries, benchmark::Counter::kIsIterationInvariantRate);
 }
 
+template <typename ExecutionSpace, class TreeType>
+void BM_radius_callback_search_kernel_query(benchmark::State &state,
```

Comment:
I'd like to see the results before seeing if we need it. As all the threads are launched at the same time in the batched mode anyway, I don't see how this is different from the usual query, and what insight it gives us. If the results are not the same, it is a problem that needs to be fixed, as imho they should be.

Reply:
You can find results above. At this point, the only difference is in predicate sorting.

Reply:
> You can find results above. At this point, the only difference is in predicate sorting.

Sure, but that's what I'm asking. If we do the same thing, i.e. sort predicates, is there any difference?

Reply:
A100, without sorting
```
BM_radius_callback_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/0/2/manual_time_median                        2538 us         7350 us           15 rate=3.94059M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/0/2/manual_time_median           2538 us         7351 us           15 rate=3.93997M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/0/2/manual_time_median                      2708 us         7516 us           15 rate=36.927M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/0/2/manual_time_median         2708 us         7516 us           15 rate=36.9247M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/0/2/manual_time_median                   13391 us        18197 us           15 rate=74.6772M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/0/2/manual_time_median      13373 us        18185 us           15 rate=74.7793M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/1/3/manual_time_median                        2485 us         7297 us           15 rate=4.02472M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/10000/10000/10/0/0/1/3/manual_time_median           2485 us         7297 us           15 rate=4.02467M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/1/3/manual_time_median                      2567 us         7375 us           15 rate=38.9593M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/100000/100000/10/0/0/1/3/manual_time_median         2567 us         7375 us           15 rate=38.9629M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/1/3/manual_time_median                    3628 us         8441 us           15 rate=275.632M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/1000000/1000000/10/0/0/1/3/manual_time_median       3627 us         8440 us           15 rate=275.685M/s
```
A100 with sorting
```
BM_radius_callback_search<ArborX::BVH<Cuda>>/10000/10000/10/1/0/0/2/manual_time_median                         252 us          279 us           15 rate=39.6682M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/10000/10000/10/1/0/0/2/manual_time_median            146 us          173 us           15 rate=68.3766M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/100000/100000/10/1/0/0/2/manual_time_median                       537 us          570 us           15 rate=186.177M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/100000/100000/10/1/0/0/2/manual_time_median          317 us          345 us           15 rate=315.853M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/0/0/2/manual_time_median                    2584 us         2844 us           15 rate=387.052M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/0/0/2/manual_time_median       4881 us         5126 us           15 rate=204.894M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/10000/10000/10/1/0/1/3/manual_time_median                        5710 us        11246 us           15 rate=1.75143M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/10000/10000/10/1/0/1/3/manual_time_median           2860 us         8412 us           15 rate=3.49685M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/100000/100000/10/1/0/1/3/manual_time_median                      5819 us        11217 us           15 rate=17.1852M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/100000/100000/10/1/0/1/3/manual_time_median         2888 us         8330 us           15 rate=34.6241M/s
BM_radius_callback_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/0/1/3/manual_time_median                    1368 us         2029 us           15 rate=730.801M/s
BM_radius_callback_kernel_query_search<ArborX::BVH<Cuda>>/1000000/1000000/10/1/0/1/3/manual_time_median       1534 us         2202 us           15 rate=651.871M/s
```

Reply:
> A100 with sorting

When you say sorting, are they sorted only in the batched mode and not in the query mode?

I assume no. The first (unsorted) clearly indicates there's no difference, which is to be expected. If that's also true for both sorted, I see no point in modifying the benchmark at all.

Reply:
Sorting should not have an impact on the batch-mode so the current performance testing setting allows us to compare sorted and unsorted results. Ultimately, I can duplicate any other test so that we have something before #902 and I chose the benchmark since the question about difference in performance came up.

Reply:
Comparing sorted and unsorted results is outside of the scope of this PR. We don't have a good handle on when sorting actually is good or not. See, for example, #474. The point of this PR is simply a refactoring and a new interface, in my opinion. Performance of the unsorted queries is completely orthogonal to that.

Reply:
Right, but we don't monitor performance in the case of not sorting so we would kill two birds with one stone.
Let me duplicate another test in `tstQueryTreeCallbacks.cpp` anyway.

Reply:
I think the only test that is required is the verification test that the results produced by a batched query and the individual calls in the loop are the same. It would have to use `TraversalPolicy` to guarantee that the sorting of the queries does not depend on the default.